src/dragonball/src/address_space_manager.rs

```diff
@@ -0,0 +1,865 @@
+    ) -> Result<Arc<GuestRegionImpl>> {
+        // Special check for 32bit host with 64bit virtual machines.
+        if region.len() > usize::MAX as u64 {
+            return Err(AddressManagerError::InvalidAddressRange(
+                region.start_addr().raw_value(),
+                region.len(),
+            ));
+        }
+        // The device MMIO regions may not be backed by memory files, so refuse to mmap them.
+        if region.region_type() == AddressSpaceRegionType::DeviceMemory {
+            return Err(AddressManagerError::InvalidOperation);
+        }
+
+        // The GuestRegionMmap/MmapRegion will take ownership of the FileOffset object,
+        // so we have to duplicate the fd here. It's really a dirty design.
+        let file_offset = match region.file_offset().as_ref() {
+            Some(fo) => {
+                let fd = dup(fo.file().as_raw_fd()).map_err(AddressManagerError::DupFd)?;
+                // Safe because we have just duplicated the raw fd.
+                let file = unsafe { File::from_raw_fd(fd) };
+                let file_offset = FileOffset::new(file, fo.start());
+                Some(file_offset)
+            }
+            None => None,
+        };
+        let perm_flags = if (region.perm_flags() & libc::MAP_POPULATE) != 0 && region.is_hugepage()
+        {
+            // mmap(MAP_POPULATE) conflicts with madive(MADV_HUGEPAGE) because mmap(MAP_POPULATE)
+            // will pre-fault in all memory with normal pages before madive(MADV_HUGEPAGE) gets
+            // called. So remove the MAP_POPULATE flag and memory will be faulted in by working
+            // threads.
+            region.perm_flags() & (!libc::MAP_POPULATE)
+        } else {
+            region.perm_flags()
+        };
+        let mmap_reg = MmapRegion::build(
+            file_offset,
+            region.len() as usize,
+            libc::PROT_READ | libc::PROT_WRITE,
+            perm_flags,
+        )
+        .map_err(AddressManagerError::MmapGuestMemory)?;
+
+        if region.is_anonpage() {
+            self.configure_anon_mem(&mmap_reg)?;
+        }
+        if let Some(node_id) = region.host_numa_node_id() {
+            self.configure_numa(&mmap_reg, node_id)?;
+        }
+        if region.is_hugepage() {
+            self.configure_thp_and_prealloc(&region, &mmap_reg)?;
+        }
+
+        let reg = GuestRegionImpl::new(mmap_reg, region.start_addr())
+            .map_err(AddressManagerError::CreateGuestMemory)?;
+        Ok(Arc::new(reg))
+    }
+
+    fn configure_anon_mem(&self, mmap_reg: &MmapRegion) -> Result<()> {
+        unsafe {
+            mman::madvise(
+                mmap_reg.as_ptr() as *mut libc::c_void,
+                mmap_reg.size(),
+                mman::MmapAdvise::MADV_DONTFORK,
+            )
+        }
+        .map_err(AddressManagerError::Madvise)
+    }
+
+    fn configure_numa(&self, mmap_reg: &MmapRegion, node_id: u32) -> Result<()> {
+        let nodemask = (1 << node_id) as u64;
```

Comment:
Again, this shift _could_ potentially lead to a crash, so safer to use `checked_shl()` (and add a test to assert that the method returns an error if node_id is too big.